hutch_python/env_version.py

```diff
@@ -0,0 +1,73 @@
+"""
+Utilities for getting relevant environment information.
+"""
+from __future__ import annotations
+
+import logging
+import os
+import os.path
+
+import pkg_resources
+
+logger = logging.getLogger(__name__)
+
+_dev_ignore_list = ['x86_64-rhel7-opt', 'x86_64-linux-opt']
+
+
+def log_env() -> None:
+    """Collect environment information and log at appropriate levels."""
+    dev_pkgs = get_standard_dev_pkgs()
+    if dev_pkgs:
+        logger.info(
+            'Using conda env %s with dev packages %s',
+            get_conda_env_name(),
+            ', '.join(sorted(dev_pkgs)),
+        )
+    else:
+        logger.info(
+            'Using conda env %s with no dev packages',
+            get_conda_env_name(),
+        )
+    logger.debug(dump_env())
+
+
+def dump_env() -> list[str]:
+    """
+    Get the full env spec.
+    conda list is slow, use pkg_resources instead
+    this might miss dev overrides
+    """
+    return sorted(str(pkg) for pkg in pkg_resources.working_set)
+
+
+def get_conda_env_name() -> str:
+    """Get the name of the conda env, or empty string if none."""
+    env_dir = os.environ.get('CONDA_PREFIX', '')
+    return os.path.basename(env_dir)
+
+
+def get_standard_dev_pkgs() -> set[str]:
+    """Check the standard dev package locations for hutch-python"""
+    pythonpath = os.environ.get('PYTHONPATH', '')
+    if not pythonpath:
+        return set()
+    pkg_names = set()
+    for part in pythonpath.split(':'):
+        if any([(s in part) for s in _dev_ignore_list]) or (not part):
```

Comment:
Unnecessary list comprehension:
```suggestion
        if any(((s in part) for s in _dev_ignore_list)) or (not part):
```

Reply:
Is the concern that we don't want to use a mutable list where we don't have to?  Apparently list comprehensions are faster than tuple comprehensions (at least in my timing tests).

Within this function there's a minimal speed difference between tuple and list comprehensions.

Reply:
It's entirely negligible as far as speed, and mutability isn't relevant. It's rather just a generally good pattern to follow.

What is the actual difference then? A temporary list is generated in the original code. In my suggestion, it's only a generator which works on an item-by-item basis - no tuple is generated, even though it may look like that syntactically. Make more sense? (Replying on mobile at the moment, otherwise I'd throw in some code as well)

I should find a better way of marking such suggestions - but laziness/brevity won out for my comment.

Reply:
Thanks for the clarification, makes perfect sense.